snapshot/snapshotfs/repofs.go

```diff
@@ -223,6 +223,54 @@ func SnapshotRoot(rep repo.Repository, man *snapshot.Manifest) (fs.Entry, error)
 	return EntryFromDirEntry(rep, man.RootEntry)
 }
 
+// AutoDetectEntryFromObjectID returns fs.Entry (either file or directory) for the provided object ID.
+// It uses heuristics to determine whether object ID is possibly a directory and treats it as such.
+func AutoDetectEntryFromObjectID(ctx context.Context, rep repo.Repository, oid object.ID, maybeName string) fs.Entry {
+	if isPossiblyDirectoryID(oid) {
+		dirEntry := DirectoryEntry(rep, oid, nil)
+		if _, err := dirEntry.Readdir(ctx); err == nil {
+			log(ctx).Debugf("%v auto-detected as directory", oid)
+			return dirEntry
+		}
+	}
+
+	if maybeName == "" {
+		maybeName = "file"
+	}
+
+	var fileSize int64
+
+	r, err := rep.OpenObject(ctx, oid)
+	if err == nil {
+		fileSize = r.Length()
+		r.Close() //nolint:errcheck
+	}
+
+	log(ctx).Debugf("%v auto-detected as a file with name %v and size %v", oid, maybeName, fileSize)
+
+	f, _ := EntryFromDirEntry(rep, &snapshot.DirEntry{
+		Name:        maybeName,
+		Permissions: 0o644, //nolint:gomnd
+		Type:        snapshot.EntryTypeFile,
+		ObjectID:    oid,
+		FileSize:    fileSize,
+	})
+
+	return f
+}
+
+func isPossiblyDirectoryID(oid object.ID) bool {
+	if ndx, ok := oid.IndexObjectID(); ok {
+		return isPossiblyDirectoryID(ndx)
+	}
+
+	if cid, _, ok := oid.ContentID(); ok {
+		return cid.Prefix() == "k"
```

Comment:
nit: Should the _well-known prefixes_ be `const`?

Reply:
done